### ISSUE TITLE
SUBSCRIBE_DONE and terminal state

### ISSUE THREAD

**Issue by kixelated** [moq-wg/moq-transport#424]
**tl;dr** when does a MoqTransport subscriber signal SUBSCRIBE_DONE to the application or clean up subscription state?

I ran into a bug when refactoring my relay to better support error conditions.

For subscriptions to the catalog or init track, my publisher would send a single OBJECT followed by an immediate SUBSCRIBE_DONE, since updates are (currently) not supported. It then removes the ID from the lookup table since this state is terminal.

Upon receiving the SUBSCRIBE_DONE, my subscriber also removes the ID from the lookup table and returns the error to the application. However, this causes a race, as the OBJECTs in flight (on another stream) will be dropped since they reference a newly non-existent subscription. This occurs the vast majority of the time especially because the control stream is higher priority.

Since SUBSCRIBE_DONE contains a final group/object, the intended behavior seems to be for the subscriber to delay until the final object is received.

But what if that object is dropped? What about the rest of the group? What happens to any other objects that might be queued (lower priority) or awaiting a retransmission?

Could the subscriber wait until all objects within subscribe range are received instead? This seems useful especially for something like FETCH, however it's currently impossible for a few reasons:

 - group_ids are not sequential and there can be intentional gaps
 - objects can be dropped, causing gaps with a group 
 - the latest object ID per group is unknown when using a stream per group.

Reliable live actually has the same exact issues. To generalize, its currently not possible to wait for all objects to arrive over the network.

**Comment by fluffy** [comment]
Imagine you had a relay with a million clients downstream that sequentially connected one at a time.  It seems like having the publisher immediately send the subcribe_done is would cause it to get a million subscribes when and send the catalog a million times, if it delayed sending the subscribe_doe that, it could get 1 from the relay and the relay could take care of the fan out.

**Comment by ianswett** [comment]
I think the subscriber needs to wait for everything up to the final Object specified in SUBSCRIBE_DONE if you want to be sure to receive everything.

If we require Object Status be used to indicate end of Group/Track and Objects that are permanently missing, then we either need a mechanism like OBJECTS_DROPPED or another Status to indicate the Object won't be delivered, or we infer that any missing Objects were dropped due to congestion/etc.

Another alternative would be to make the publisher completely deliver all Streams/Datagrams with Objects on them prior to sending SUBSCRIBE_DONE.  In that case, the subscriber would have received everything it was going to get prior to SUBSCRIBE_DONE.  I'm unsure if this is practical for all QUIC and WebTransport implementations?

**Comment by ianswett** [comment]
I believe this would be fixed by #609
